### PR TITLE
Axiom constraints 4

### PR DESCRIPTION
**Pull request by andreiburdusa**
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 1]
```diff
@@ -112,7 +112,7 @@ termEquals
     -> TermLike variable
     -> MaybeT simplifier (OrPredicate variable)
 termEquals first second = MaybeT $ do
-    maybeResults <- BranchT.gather $ runMaybeT $ termEqualsAnd first second
+    maybeResults <- BranchT.gather $ runMaybeT $ termEqualsAnd Predicate.top first second
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Simplification/AndTerms.hs]: Line lengths over 80 characters are discouraged by the style guide:
https://github.com/kframework/kore/wiki/Coding-Standards
We go over that limit when it significantly helps readability, e.g. in some tests and in the `andEqualsFunctions` implementation below.

**Comment by virgil-serbanuta** [comment]
LGTM, but I'll let ttuegel approve.

File: kore/src/Kore/Repl/State.hs [hunk 2]
```diff
@@ -459,7 +459,7 @@ runUnifier first second = do
     mvar <- asks logger
     liftSimplifierWithLogger mvar
         . runUnifierWithExplanation
-        $ unifier first second
+        $ unifier IPredicate.top first second
```
[the remark this answers is not in the document]
Comment (andreiburdusa) [reply kore/src/Kore/Repl/State.hs]: We changed `unificationProcedure` from Procedure.hs so we had to change `data Config claim m` which contains a unifier

File: kore/test/Test/Kore/Step/MockSimplifiers.hs [hunk 3]
```diff
@@ -19,7 +19,8 @@ substitutionSimplifier =
   where
     -- TODO (thomas.tuegel): Remove simplifierWrapPredicate.
     simplifierWrapPredicate =
-        termLikeSimplifier $ \patt ->
+        termLikeSimplifier $ \predicate patt -> (flip const) predicate $
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/MockSimplifiers.hs]: Why not?

```suggestion
        termLikeSimplifier $ \ _ patt ->
```

File: kore/src/Kore/Step/Step.hs [hunk 4]
```diff
@@ -227,8 +228,9 @@ unifyRule
     -- configuration.
     let
         RulePattern { left = ruleLeft } = rule'
-    unification <- unifyPatterns ruleLeft initialTerm
-    -- Combine the unification solution with the rule's requirement clause.
+    unification <- unifyPatterns (withoutTerm initial) ruleLeft initialTerm
+    -- Combine the unification solution with the rule's requirement clause,
+    -- unless it's rendundant to do so.
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: This looks like a merge conflict; the redundancy check was removed below. Could you fix the comment?

File: kore/src/Kore/Unification/Procedure.hs [hunk 5]
```diff
@@ -52,10 +52,11 @@ unificationProcedure
         , FreshVariable variable
         , MonadUnify unifier
         )
-    => TermLike variable
+    => Predicate variable
+    -> TermLike variable
     -> TermLike variable
     -> unifier (Predicate variable)
-unificationProcedure  p1 p2
+unificationProcedure predicate p1 p2
```
Comment (ttuegel) [inline kore/src/Kore/Unification/Procedure.hs]: `unificationProcedure` creates a configuration predicate, so we shouldn't need to take one as an argument.

File: kore/src/Kore/Unification/Procedure.hs [hunk 6]
```diff
@@ -77,7 +78,7 @@ unificationProcedure  p1 p2
     if Conditional.isBottom pat
         then empty
         else do
-            orCeil <- Ceil.makeEvaluateTerm Predicate.topTODO term
+            orCeil <- Ceil.makeEvaluateTerm predicate term
```
Comment (ttuegel) [inline kore/src/Kore/Unification/Procedure.hs]: Use the predicate which results from unification:
```suggestion
            orCeil <- Ceil.makeEvaluateTerm (Pattern.withoutTerm pat) term
```

**Pull request merged**